### PR TITLE
[WIP] Genetic Algorithm for Adders Configuration

### PR DESCRIPTION
**Pull request by sdamghan**
#### Description
Using the Genetic Algorithm to find the best configuration of adders in the netlist.

#### Types of changes

- [ ] Bug fix (change which fixes an issue)
- [x] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

#### Checklist:

- [x] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed

### CODE REVIEW

File: ODIN_II/SRC/soft_logic_def_parser.cpp [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+soft_sub_structure *fetch_blk(std::string op, int width)
+{
+	std::string lookup = op + "_" + std::to_string(width);
+
+	auto candidate = soft_def_map.find(lookup);
+	if(candidate == soft_def_map.end())
+		return NULL;
+
+	return soft_def_map[lookup];
+}
```
Comment (jeanlego) [inline ODIN_II/SRC/soft_logic_def_parser.cpp]: drop this file since you don't need to parse an input file for the adder definition

File: ODIN_II/SRC/partial_map.cpp [hunk 2]
```diff
@@ -1170,3 +1207,265 @@ void instantiate_arithmetic_shift_right(nnode_t *node, short mark, netlist_t *ne
...
+	{
+		int point = rand()%num_of_adders;
+		new_chromosome[point] = rand()%3;
+
+		num_of_genes_will_be_changed--;
+	}
+	
+	return new_chromosome;
+}
+//MEHRSHAD//
```
Comment (jeanlego) [inline ODIN_II/SRC/partial_map.cpp]: move these functions into ga_adder.cpp, put the function headers into a ga_adder.hpp file
remove your name from the comments.

File: ODIN_II/SRC/partial_map.cpp [hunk 3]
```diff
@@ -190,7 +207,27 @@ void partial_map_node(nnode_t *node, short traverse_number, netlist_t *netlist)
 				// Check if the size of this adder is greater than the hard vs soft logic threshold
 					instantiate_hard_adder(node, traverse_number, netlist);
 			}else{
-				instantiate_add_w_carry(node, traverse_number, netlist);
+				//MEHRSHAD//
```
Comment (jeanlego) [inline ODIN_II/SRC/partial_map.cpp]: remove your name

File: ODIN_II/SRC/partial_map.cpp [hunk 4]
```diff
@@ -190,7 +207,27 @@ void partial_map_node(nnode_t *node, short traverse_number, netlist_t *netlist)
...
+				adders_list[num_of_adders]->input->count = node->num_input_pins;
+				adders_list[num_of_adders]->input->is_adder = 'y';
+				
+				adders_list[num_of_adders]->output->pins = node->output_pins;
+				adders_list[num_of_adders]->output->count = node->num_output_pins;
+				adders_list[num_of_adders]->output->is_adder = 'y';
+
+				chromosome[num_of_adders] = adders_list[num_of_adders]->type;
+				num_of_adders++;
+				//MEHRSHAD//
```
Comment (jeanlego) [inline ODIN_II/SRC/partial_map.cpp]: remove your name

File: ODIN_II/SRC/partial_map.cpp [hunk 5]
```diff
@@ -42,6 +42,23 @@ OTHER DEALINGS IN THE SOFTWARE.
...
+npin_t **cloud_pins_list;
+nnet_t  **cloud_nets_list;
+nnode_t **cloud_nodes_list;
+
+int num_cloud_pins;
+int num_cloud_nets;
+int num_cloud_nodes;
+
+long num_of_adders = 0;
+// MEHRSHAD //
```
Comment (jeanlego) [inline ODIN_II/SRC/partial_map.cpp]: remove your name

File: ODIN_II/SRC/odin_ii.cpp [hunk 6]
```diff
@@ -581,15 +589,6 @@ void get_options(int argc, char** argv) {
...
 
-	//adjust thread count
-	int thread_requested = global_args.parralelized_simulation;
-	int max_thread = std::thread::hardware_concurrency();
-
-	global_args.parralelized_simulation.set(
-		std::max(1, std::min( thread_requested, std::min( (CONCURENCY_LIMIT-1) , max_thread )))
-		,argparse::Provenance::SPECIFIED
-	);
-
```
Comment (jeanlego) [inline ODIN_II/SRC/odin_ii.cpp]: why is this removed?

File: ODIN_II/SRC/odin_ii.cpp [hunk 7]
```diff
@@ -162,6 +166,10 @@ static ODIN_ERROR_CODE synthesize_verilog()
 	printf("Performing Partial Map to target device\n");
 	partial_map_top(verilog_netlist);
 
+	// MEHRSHAD //
+	partial_map_adders_GA_top(verilog_netlist);
```
Comment (jeanlego) [inline ODIN_II/SRC/odin_ii.cpp]: we use snake case so avoid capitalizing

File: ODIN_II/SRC/netlist_utils.cpp [hunk 8]
```diff
@@ -1428,11 +1428,4 @@ void remove_fanout_pins_from_net(nnet_t *net, npin_t * /*pin*/, int id)
 // 	printf("\t related net : %s related node: %s\n",pin->net->name,pin->node->name);
 //  	}
 // }
-// }
```
Comment (jeanlego) [inline ODIN_II/SRC/netlist_utils.cpp:1431]: avoid whitespace changes

File: ODIN_II/SRC/netlist_utils.cpp [hunk 9]
```diff
@@ -245,7 +245,7 @@ npin_t* free_npin(npin_t* to_free) {
  *  Should only be called by the parent functions,
  *  copy_input_npin & copy_output_npin
  *-----------------------------------------------------------------------*/
-static npin_t *copy_npin(npin_t *copy_pin)
```
Comment (jeanlego) [inline ODIN_II/SRC/netlist_utils.cpp:248]: functions are declared static if you don't expect to use them outside the file, this way you can reuse the name more than once and it can avoid weird errors. also, it reduces code size since static function don't need to be linked.

File: ODIN_II/SRC/include/soft_logic_def_parser.h [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+  char *type;
+	char *name;
+	int bitsize;
+};
+
+void read_soft_def_file(t_model *hard_adder_models);
+soft_sub_structure *fetch_blk(std::string, int width);
+
+
+#endif
```
Comment (jeanlego) [inline ODIN_II/SRC/include/soft_logic_def_parser.h]: remove this file

File: ODIN_II/SRC/include/odin_types.h [hunk 11]
```diff
@@ -101,8 +104,8 @@ struct global_args_t
     argparse::ArgValue<bool> all_warnings;
     argparse::ArgValue<bool> show_help;
 
-	argparse::ArgValue<std::string> adder_def; //DEPRECATED
-
+	argparse::ArgValue<std::string> adder_def; //carry skip adder skip size
```
Comment (jeanlego) [inline ODIN_II/SRC/include/odin_types.h]: don't use it this way. this expects either default, optimized or a file name. Make a new one and use a boolean to enable GA adder

File: ODIN_II/SRC/include/netlist_utils.h [hunk 12]
```diff
@@ -13,6 +13,9 @@ npin_t* allocate_npin();
 npin_t *get_zero_pin(netlist_t *netlist);
 npin_t *get_pad_pin(netlist_t *netlist);
 npin_t *get_one_pin(netlist_t *netlist);
+//MEHRSHAD//
```
Comment (jeanlego) [inline ODIN_II/SRC/include/netlist_utils.h]: ? is this necessary

File: ODIN_II/SRC/adders.cpp [hunk 13]
```diff
@@ -30,6 +30,7 @@ OTHER DEALINGS IN THE SOFTWARE.
 #include "odin_types.h"
 #include "odin_util.h"
 #include "node_creation_library.h"
+#include "soft_logic_def_parser.h"
```
Comment (jeanlego) [inline ODIN_II/SRC/adders.cpp]: remove this header

File: ODIN_II/SRC/adders.cpp [hunk 14]
```diff
@@ -1483,21 +1562,135 @@ static nnode_t *make_adder(operation_list funct, nnode_t *current_adder, nnode_t
...
+	nnode_t *initial_carry 		= (subtraction)? netlist->vcc_node: netlist->gnd_node;
+	for(int start_pin=0, current_counter=1 ; start_pin<width[0]; start_pin+=blk_size, current_counter++)
 	{
-		/* set of flags for building purposes */
-		short construct_last_carry_flag = (i != width[0]-1 || !subtraction)? 1:0;
+		std::string my_type 			= "soft";
+		std::string sub_structure = "default";
+		blk_size 									= width[0]-start_pin;
+
+		soft_sub_structure *def = fetch_blk("+",width[0]-start_pin);
```
Comment (jeanlego) [inline ODIN_II/SRC/adders.cpp]: remove this line and fix all the dependant lines bellow.

File: ODIN_II/SRC/include/netlist_utils.h (hunk 12 above)
Comment (sdamghan) [reply ODIN_II/SRC/include/netlist_utils.h]: I need this function in the ga_adder.cpp (GA functions at the end of partial_map.cpp), so in the other ways, I can not use it there!

File: ODIN_II/SRC/netlist_utils.cpp (hunk 9 above)
Comment (sdamghan) [reply ODIN_II/SRC/netlist_utils.cpp:248]: I need this function in the ga_adder.cpp (GA functions at the end of partial_map.cpp), so in the other ways, I can not use it there!
